homeassistant/components/weather/smhi.py

```diff
@@ -0,0 +1,234 @@
+    'pouring': [10, 20],
+    'rainy': [8, 9, 18, 19],
+    'snowy': [15, 16, 17, 25, 26, 27],
+    'snowy-rainy': [12, 13, 14, 22, 23, 24],
+    'sunny': [1, 2],
+    'windy': [],
+    'windy-variant': [],
+    'exceptional': [],
+}
+
+
+# 5 minutes between retrying connect to API again
+RETRY_TIMEOUT = 5*60
+
+MIN_TIME_BETWEEN_UPDATES = timedelta(minutes=31)
+
+
+async def async_setup_platform(hass, config, async_add_entities,
+                               discovery_info=None):
+    """Old way of setting up components.
+
+    Can only be called when a user accidentally mentions smhi in their
+    config. But even in that case it would have been ignored.
+    """
+    pass
+
+
+async def async_setup_entry(hass: HomeAssistant,
+                            config_entry: ConfigEntry,
+                            config_entries) -> bool:
+    """Add a weather entity from map location."""
+    location = config_entry.data
+    name = location[CONF_NAME]
+
+    entity = SmhiWeather(name, location[CONF_LATITUDE],
+                         location[CONF_LONGITUDE])
+    entity.entity_id = ENTITY_ID_SENSOR_FORMAT.format(name)
+
+    config_entries([entity], True)
+    return True
+
+
+class SmhiWeather(WeatherEntity):
+    """Representation of a weather entity."""
+
+    _forecasts = None
+
+    def __init__(self, name: str, latitude: str,
+                 longitude: str) -> None:
+        """Initialize the Smhi weather entity."""
+        self._name = name
+        self._latitude = latitude
+        self._longitude = longitude
+        self._forecasts = None
+
+    # pylint: disable=broad-except
+    @Throttle(MIN_TIME_BETWEEN_UPDATES)
+    async def async_update(self, *_, counter=0):
+        """Refresh the forecast data from smhi weather API."""
+        if counter > 2:
+            # Failed 3 times, 15 minutes. Give up!
+            return
+
+        try:
+            with async_timeout.timeout(10, loop=self.hass.loop):
+                self._forecasts = await self.get_weather_forecast()
+
+        except asyncio.TimeoutError:
+            _LOGGER.exception("Failed to connect to SMHI API, "
+                              "timeout, retry in 5 minutes")
+            self.hass.helpers.event.async_call_later(
+                RETRY_TIMEOUT, self.retry_update(counter+1))
+        except Exception:
+            _LOGGER.exception("Failed to connect to SMHI API, "
+                              "retry in 5 minutes")
+            self.hass.helpers.event.async_call_later(
+                RETRY_TIMEOUT, self.retry_update(counter+1))
+
+    async def retry_update(self, counter: int):
+        """Retry refresh weather forecast."""
+        self.async_update(counter=counter)
+
+    async def get_weather_forecast(self) -> []:
+        """Return the current forecasts from SMHI API."""
+        from smhi import Smhi
+
+        smhi_api = Smhi(self._longitude, self._latitude)
+        websession = aiohttp_client.async_get_clientsession(self.hass)
```

Comment:
Is it safe to store the websession? I thought it was per request thing. I could definitely store the API object

Reply:
The aiohttp session is preferably kept during the whole lifetime of the app. So it should be safe to store it. Normally it is stored on the client that uses it.

See the note here:
https://aiohttp.readthedocs.io/en/stable/client_quickstart.html

Reply:
Store it in init, but we can instantiate the api before and pass it in to the entity.

For tests probably make a pytest fixture that mocks out the client api, which adds very little overhead when writing the tests.

Reply:
Seems like it fails when doing the `self._websession = aiohttp_client.async_get_clientsession(self.hass)` in __init__ cause hass attribute is probably not assigned automatically yet. And since I removed the hass input to the constructor it fails.

Reply:
Yes, so instead create the api instance in entry set up and pass it it to the entity.

Reply:
Regarding the session, the best solution would be to update the library to have the api client accept an aiohttp session as a keyword argument. Then we can pass the our home assistant session to the client instance, which saves it, and we save the client on the entity.

Reply:
Ofc thats the solution. Awesome!